source/src/localisation/cs_TrackableParser.cpp

```diff
@@ -124,136 +223,371 @@ void TrackableParser::handleScannedDevice(scanned_device_t* device) {
 
-	if (parsingFilterChunk_ptr + cmd_data->chunkSize > _filterBuffer + FILTER_BUFFER_SIZE) {
-		// chunk would overwrite end of filterbuffer.
-		// (Unlikely to happen, previous guards should have caught this.)
-		return ERR_NO_SPACE;
-	}
-
 	// apply filter chunk, counting chunk index from metadata onwards:
 	std::memcpy(parsingFilterChunk_ptr, cmd_data->chunk, cmd_data->chunkSize);
 
 	return ERR_SUCCESS;
 }
 
-cs_ret_code_t TrackableParser::handleRemoveFilterCommand(
-		trackable_parser_cmd_remove_filter_t* cmd_data) {
-	// TODO(Arend): implement later.
-	return ERR_NOT_IMPLEMENTED;
+cs_ret_code_t TrackableParser::handleRemoveFilterCommand(trackable_parser_cmd_remove_filter_t* cmd_data) {
+	startProgress();
+
+	return deallocateParsingFilter(cmd_data->filterId) ? ERR_SUCCESS : ERR_NOT_FOUND;
 }
 
 cs_ret_code_t TrackableParser::handleCommitFilterChangesCommand(
 		trackable_parser_cmd_commit_filter_changes_t* cmd_data) {
-	// TODO(Arend): implement later.
-	// - compute and check all filter sizes
-	// - compute and check all filter crcs
-	// - compute and check(?) master crc
-	// - persist all filters
-	// - unset in progress flag (async?)
-	// - broadcast update to the mesh
-	return ERR_NOT_IMPLEMENTED;
+
+	if (checkFilterSizeConsistency()) {
+		return ERR_WRONG_STATE;
+	}
+
+	computeCrcs();
+
+	uint16_t masterHash = masterCrc();
+	if (cmd_data->masterCrc != masterHash) {
+		LOGTrackableParserWarn("Master Crc did not match (%x != %x, filterModificationInProgress stays true.",
+				cmd_data->masterCrc,
+				masterHash);
+		return ERR_MISMATCH;
+	}
+
+	// finish commit by writing the _masterHash
+	_masterHash                  = masterHash;
+	filterModificationInProgress = false;
+
+	// NOTE: if writing flash, this may need to be delayed until write success.
+	return ERR_SUCCESS;
 }
 
-cs_ret_code_t TrackableParser::handleGetFilterSummariesCommand(
-		trackable_parser_cmd_get_filer_summaries_t* cmd_data) {
-	// TODO(Arend): implement later.
-	return ERR_NOT_IMPLEMENTED;
+void TrackableParser::handleGetFilterSummariesCommand(
+		trackable_parser_cmd_get_filter_summaries_t* cmd_data, cs_result_t& result) {
+	LOGTrackableParserDebug("handle get filter summaries:");
+	for (auto& trackingFilter : _parsingFilters) {
+		if (trackingFilter == nullptr) {
+			LOGTrackableParserDebug("filter: nullptr");
+			continue;
+		}
+		LOGTrackableParserDebug("filter: %u", trackingFilter->runtimedata.filterId);
+	}
+
+	// stack allocate a buffer summaries object fitting at most max summaries:
+	auto requiredBuffSize = sizeof(trackable_parser_cmd_get_filter_summaries_ret_t)
+							+ sizeof(tracking_filter_summary_t) * _parsingFiltersCount;
+
+	if (result.buf.len < requiredBuffSize) {
+		result.returnCode = ERR_BUFFER_TOO_SMALL;
+		return;
+	}
+
+	// placement new constructs the object in the buff and now has enough space for the summaries.
+	auto retvalptr  = new (result.buf.data) trackable_parser_cmd_get_filter_summaries_ret_t;
```

Comment:
What does this do?
Why not just cast the buffer?

Reply:
This is the most type safe version of a cast you can make: placement new.

It ensures any default values will be set and possibly calls required constructors, as if it was a normal call to `new`, but instead of the heap it uses the given pointer `result.buf.data`. The returned object `retvalptr` is of type `trackable_parser..._ret_t*`.